### PR TITLE
Featured books

### PR DESCRIPTION
**Pull request by sisiz**
Creating featured book api to add a book to the add books to the featured books list on CNX and upload cover images

### CODE REVIEW

**Comment by karenc** [comment]
@sisiz this is great!  We've been waiting for an admin interface to edit the featured books.

Could you structure the commits so that each commit contains one logical change please?  It is pretty difficult to review as there are so many small commits in this pull request.

There are lots of blog posts and wiki pages on git commits to help you with this:

 - https://github.com/openstax/napkin-notes/blob/master/CONTRIBUTING.md#use-a-good-commit-message
 - https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

Having good commits is going to help us review, debug, revert, move our code easier.

**Comment by sisiz** [comment]
@karenc No problem, I tried to rebase the commits down. there are a few things that were fixes of a previous thing commit that I only discovered after starting the next part so the change is with the wrong commit message. does that make sense? But mostly it matches the commit message

**Comment by karenc** [comment]
@sisiz The first commit looks good, the rest of the commits seem to me like they fix the first commit.  So you can probably squash them later.

About the first commit message, I think the subject line (the first line) is good, but it is almost always necessary to add more details to what you're doing.  For example, have a look at https://wiki.openstack.org/wiki/GitCommitMessages#Example_1_4.  Things we should try to include in the commit messages are:

 - WHY? what the original problem is
 - WHAT? the functional change being made
 - what the result of the change is
 - (optional) scope for future improvement

**Comment by sisiz** [comment]
@karenc should I just mush all the commits together with git rebase and baking the commits after the first just fix-ups? I'll try to work on making the commit messages better.

**Comment by karenc** [comment]
Yeah I think for this PR, all the changes belong to the same feature and it's not too big so it's ok to squash everything into one commit.

File: cnxpublishing/views/admin.py [hunk 1]
```diff
@@ -522,3 +526,169 @@ def admin_content_status_single_POST(request):
...
+@view_config(route_name='admin-featured-books', request_method='GET',
+             renderer='templates/featured-books.html',
+             permission='view')
+def admin_featured_books(request):
+    settings = request.registry.settings
+    with db_connect() as db_conn:
+        with db_conn.cursor() as cursor:
+            cursor.execute("""
+                SELECT ident_hash(f.uuid, f.major_version, f.minor_version),
+                (SELECT name
```
Comment (reedstrm) [inline cnxpublishing/views/admin.py]: This repeats logic that's in cnx-db/archive-sql. Could imported here.

File: cnxpublishing/views/admin.py [hunk 2]
```diff
@@ -522,3 +526,169 @@ def admin_content_status_single_POST(request):
...
+                         f.major_version is NULL)
+                    and (f.minor_version=m.minor_version OR
+                         f.minor_version is NULL)
+                ORDER by revised DESC LIMIT 1)
+                FROM featured_books as f;""")
+            data = cursor.fetchall()
+            featured_books = []
+            for row in data:
+                title = row[1]
+                if title is None:
```
Comment (reedstrm) [inline cnxpublishing/views/admin.py]: Hmm, good to play it safe, but we have a `NOT NULL` on the `name` column of `modules`, where `title` came from, so unnecessary complexity here.

File: cnxpublishing/views/admin.py [hunk 3]
```diff
@@ -522,3 +526,169 @@ def admin_content_status_single_POST(request):
...
+                # have to add the file
+                cursor.execute("""\
+                    INSERT INTO files (file, media_type)
+                    VALUES (%s, 'image/png')
+                    RETURNING fileid;
+                    """, vars=(file_bytes, ))
+                fileid = cursor.fetchall()[0][0]
+                cursor.execute("""\
+                    INSERT INTO module_files (module_ident, fileid, filename)
+                    VALUES (%s, %s, 'featured-cover.png')
```
Comment (reedstrm) [inline cnxpublishing/views/admin.py]: Instead of using magic filename, let's just keep the fileid in the featured_books table.

**Comment by sisiz** [comment]
@reedstrm I made some of the changes requested but I'm not quite sure how to make the pr use the database change in the cnx-db pr because the format of the travis file has changed a bit.

File: cnxpublishing/tests/views/test_admin.py [hunk 4]
```diff
@@ -534,3 +533,228 @@ def test_admin_content_status_single_page_POST_bad_uuid(self):
...
+        cls.settings = integration_test_settings()
+        from cnxpublishing.config import CONNECTION_STRING
+        cls.db_conn_str = cls.settings[CONNECTION_STRING]
+        cls.db_connect = staticmethod(db_connection_factory())
+
+    def setUp(self):
+        with self.db_connect() as db_conn:
+            with db_conn.cursor() as cursor:
+                cursor.execute("DROP SCHEMA public CASCADE")
+                cursor.execute("CREATE SCHEMA public")
```
[the remark this answers is not in the document]
Comment (sisiz) [reply cnxpublishing/tests/views/test_admin.py]: Some of the other test files/chases do this as well and I just copied it (test_api_keys.py ApiKeyViewsTestCase class, base.py setUpClass class, ect) but others do not (test_publishing.py PublishViewsTestCase, ect ). I'm not quite sure if it is necessary for it to be there, I can look into the code a bit more but I was having some issues without it but I was also having some issues with my set up in general at the time

**Comment by brittweinstein** [comment]
CLOSING. If this presents itself as an issue again, we will reopen.

**Pull request closed without merging**